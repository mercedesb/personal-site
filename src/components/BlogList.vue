<template>
  <transition name='fade'>
    <div v-if="posts.length" class='PageContent PageContent--wide'>
      <BlogListItem v-if="featuredPost"
        v-bind="featuredPost"
        :key="featuredPost.id"
      ></BlogListItem>
      <BlogListItem v-for="blogPost in remainingPosts"
        v-bind="blogPost"
        :key="blogPost.id"
      ></BlogListItem>
    </div>
  </transition>
</template>

<script>
import BlogListItem from './BlogListItem.vue'

export default {
  components: {
    BlogListItem
  },
  props: {
    color: String
  },
  computed: {
    posts () {
      return this.$store.state.blogPosts.map((post) => {
        return {
          id: post.id,
          color: this.color,
          title: post.title,
          preamble: post.preamble,
          urlSegment: post.urlSegment,
          date: new Date(post.publishDate)
        }
      })
    },
    featuredPost () {
      return {
        featured: true,
        ...this.posts[0]
      }
    },
    remainingPosts () {
      return this.posts.slice(1)
    }
  },
  created () {
    this.$store.dispatch('getBlogPosts')
  }
}
</script>

<style lang="scss">
</style>
