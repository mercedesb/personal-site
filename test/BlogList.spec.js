import { shallowMount, mount, createLocalVue } from '@vue/test-utils'
import BlogList from '@/components/BlogList.vue'
import Setup from './Setup'
import Vuex from 'vuex'

describe('BlogList', () => {
  let component

  Setup.configure()

  const initialProps = {
    color: 'gray'
  }

  const store = new Vuex.Store({
    state: {
      blogPosts: Setup.blogPosts
    }
  })
  store.dispatch = jest.fn()

  const shallow = propsData => shallowMount(BlogList, { 
    store,
    propsData: {
      ...initialProps,
      ...propsData
    }
   })

  describe('Snapshots', () => {
    describe('with a color passed in', () => {
      it('matches snapshot', () => {
        component = shallow()
        expect(component.element).toMatchSnapshot()
      })
    })

    describe('without a color passed in', () => {
      it('matches snapshot', () => {
        component = shallow({color: undefined})
        expect(component.element).toMatchSnapshot()
      })
    })
  })

  describe('Properties', () => {
    it('has a color property', () => {
      component = shallow()
      expect(component.props().color).toBe('gray')
    })
  })

  describe('Computed', () => {
    describe('posts', () => {
      beforeEach(() => {
        component = shallow()
      })

      it('has posts populated', () => {
        expect(component.vm.posts).toBeDefined()
      })

      it('creates a color prop on the posts', () => {
        expect(component.vm.posts[0].color).toEqual(initialProps.color)
      })

      it('creates a date prop on the posts', () => {
        expect(component.vm.posts[0].date).toBeDefined()
      })
    })

    describe('featuredPost', () => {
      it('has featuredPost populated', () => {
        expect(component.vm.featuredPost).toBeDefined()
      })

      it('returns the first post as featured', () => {
        const expected = {
          featured: true,
          ...component.vm.posts[0]
        }
        expect(component.vm.featuredPost).toEqual(expected)
      })
    })

    describe('remainingPosts', () => {
      it('has remainingPosts populated', () => {
        expect(component.vm.remainingPosts).toBeDefined()
      })

      it('returns the second through end post', () => {
        const expected = component.vm.posts.slice(1)
        expect(component.vm.remainingPosts).toEqual(expected)
      })
    })
  })

  describe('Lifecycle', () => {
    describe('created', () => {
      it('dispatches getBlogPosts to the store', () => {
        component = mount(BlogList, { 
          store,
          propsData: {
            ...initialProps
          }
         })
        expect(store.dispatch).toHaveBeenCalledWith('getBlogPosts')
      })
    })
  })
})