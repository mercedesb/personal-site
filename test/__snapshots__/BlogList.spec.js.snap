// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BlogList Snapshots with a color passed in matches snapshot 1`] = `
<div
  class="PageContent PageContent--wide"
  name="fade"
>
  <!---->
   
  <!---->
  <!---->
</div>
`;

exports[`BlogList Snapshots without a color passed in matches snapshot 1`] = `
<div
  class="PageContent PageContent--wide"
  name="fade"
>
  <!---->
   
  <!---->
  <!---->
</div>
`;
